File: javascript/chat-agent/ChatAgent.js
/**
 * Chat Agent - LLM-powered conversational agent
 *
 * Features:
 * - Multi-turn conversations
 * - Context management
 * - Message history
 * - System prompts
 */

const BaseAgent = require('../base/Agent');

class ChatAgent extends BaseAgent {
  /**
   * Initialize chat agent
   */
  async _initialize() {
    this.state.messages = [];
    this.state.systemPrompt = this.config.systemPrompt || 'You are a helpful AI assistant.';
    this.state.maxHistory = this.config.maxHistory || 20;

    // Add system message
    if (this.state.systemPrompt) {
      this.state.messages.push({
        role: 'system',
        content: this.state.systemPrompt
      });
    }
  }

  /**
   * Send a message and get a response
   * @param {string} message - User message
   * @param {Object} options - Additional options
   * @returns {Promise<string>} Agent response
   */
  async execute(message, options = {}) {
    if (!this._initialized) {
      await this.initialize();
    }

    // Add user message to history
    this.addMessage('user', message);

    // Generate response
    const response = await this._generateResponse(message, options);

    // Add assistant response to history
    this.addMessage('assistant', response);

    // Trim history if needed
    this._trimHistory();

    return response;
  }

  /**
   * Generate response using LLM provider
   * Override this method to integrate with actual LLM providers
   * @private
   */
  async _generateResponse(message, options) {
    const provider = this.config.provider || 'mock';

    if (provider === 'mock') {
      return `Mock response to: ${message}`;
    }

    // Add your LLM integration here
    // Example for OpenAI:
    // const OpenAI = require('openai');
    // const openai = new OpenAI({ apiKey: this.config.apiKey });
    // const response = await openai.chat.completions.create({
    //   model: this.config.model || 'gpt-4',
    //   messages: this.state.messages
    // });
    // return response.choices[0].message.content;

    throw new Error(`Provider '${provider}' not implemented`);
  }

  /**
   * Add a message to chat history
   * @param {string} role - Message role (user, assistant, system)
   * @param {string} content - Message content
   */
  addMessage(role, content) {
    this.state.messages.push({ role, content });
  }

  /**
   * Trim message history to max length
   * @private
   */
  _trimHistory() {
    const maxHistory = this.state.maxHistory;

    // Always keep system message (index 0)
    if (this.state.messages.length > maxHistory + 1) {
      const systemMsg = this.state.messages[0];
      this.state.messages = [
        systemMsg,
        ...this.state.messages.slice(-(maxHistory))
      ];
    }
  }

  /**
   * Clear chat history (except system message)
   */
  clearHistory() {
    const systemMsg = this.state.messages[0] || null;
    this.state.messages = systemMsg ? [systemMsg] : [];
    this.log('info', 'Chat history cleared');
  }

  /**
   * Get chat history
   * @returns {Array} Message history
   */
  getHistory() {
    return this.state.messages;
  }

  /**
   * Update system prompt
   * @param {string} prompt - New system prompt
   */
  setSystemPrompt(prompt) {
    this.state.systemPrompt = prompt;
    if (this.state.messages.length > 0 && this.state.messages[0].role === 'system') {
      this.state.messages[0].content = prompt;
    } else {
      this.state.messages.unshift({
        role: 'system',
        content: prompt
      });
    }
    this.log('info', 'System prompt updated');
  }
}

module.exports = ChatAgent;

// Example usage
if (require.main === module) {
  (async () => {
    console.log('Chat Agent Example');
    console.log('='.repeat(50));

    const agent = new ChatAgent({
      systemPrompt: 'You are a helpful coding assistant.',
      maxHistory: 10
    });

    await agent.use(async (agent) => {
      const messages = [
        "Hello! Can you help me with JavaScript?",
        "What's the difference between let and const?",
        "Thanks for the help!"
      ];

      for (const msg of messages) {
        console.log(`\nUser: ${msg}`);
        const response = await agent.execute(msg);
        console.log(`Agent: ${response}`);
      }

      console.log('\n' + '='.repeat(50));
      console.log(`Total messages in history: ${agent.getHistory().length}`);
    });
  })();
}

File: python/chat_agent/chat_agent.py
"""
Chat Agent - LLM-powered conversational agent

This agent can be integrated with various LLM providers (OpenAI, Anthropic, etc.)
"""

import sys
import os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))

from base.agent import BaseAgent
from typing import List, Dict, Any, Optional


class ChatAgent(BaseAgent):
    """
    A conversational agent that maintains chat history and context

    Features:
    - Multi-turn conversations
    - Context management
    - Message history
    - System prompts
    - Token counting (basic)

    Example:
        agent = ChatAgent({
            'system_prompt': 'You are a helpful assistant',
            'max_history': 10
        })
        agent.initialize()
        response = agent.execute('Hello, how are you?')
        print(response)
    """

    def _initialize(self) -> None:
        """Initialize chat agent"""
        self.state['messages'] = []
        self.state['system_prompt'] = self.config.get(
            'system_prompt',
            'You are a helpful AI assistant.'
        )
        self.state['max_history'] = self.config.get('max_history', 20)

        # Add system message
        if self.state['system_prompt']:
            self.state['messages'].append({
                'role': 'system',
                'content': self.state['system_prompt']
            })

    def execute(self, message: str, **kwargs) -> str:
        """
        Send a message and get a response

        Args:
            message: User message
            **kwargs: Additional parameters (provider-specific)

        Returns:
            Agent response
        """
        if not self._initialized:
            self.initialize()

        # Add user message to history
        self.add_message('user', message)

        # Generate response (this is a placeholder - integrate with actual LLM)
        response = self._generate_response(message, **kwargs)

        # Add assistant response to history
        self.add_message('assistant', response)

        # Trim history if needed
        self._trim_history()

        return response

    def _generate_response(self, message: str, **kwargs) -> str:
        """
        Generate response using LLM provider

        Override this method to integrate with actual LLM providers:
        - OpenAI API
        - Anthropic Claude
        - Local models (Ollama, etc.)
        - Any other provider
        """
        # This is a placeholder implementation
        # Replace with actual LLM call
        provider = self.config.get('provider', 'mock')

        if provider == 'mock':
            return f"Mock response to: {message}"

        # Add your LLM integration here
        # Example for OpenAI:
        # import openai
        # response = openai.ChatCompletion.create(
        #     model=self.config.get('model', 'gpt-4'),
        #     messages=self.state['messages']
        # )
        # return response.choices[0].message.content

        raise NotImplementedError(f"Provider '{provider}' not implemented")

    def add_message(self, role: str, content: str) -> None:
        """Add a message to chat history"""
        self.state['messages'].append({
            'role': role,
            'content': content
        })

    def _trim_history(self) -> None:
        """Trim message history to max_history length"""
        max_history = self.state['max_history']

        # Always keep system message (index 0)
        if len(self.state['messages']) > max_history + 1:
            system_msg = self.state['messages'][0]
            self.state['messages'] = (
                [system_msg] +
                self.state['messages'][-(max_history):]
            )

    def clear_history(self) -> None:
        """Clear chat history (except system message)"""
        system_msg = self.state['messages'][0] if self.state['messages'] else None
        self.state['messages'] = [system_msg] if system_msg else []
        self.logger.info("Chat history cleared")

    def get_history(self) -> List[Dict[str, str]]:
        """Get chat history"""
        return self.state['messages']

    def set_system_prompt(self, prompt: str) -> None:
        """Update system prompt"""
        self.state['system_prompt'] = prompt
        if self.state['messages'] and self.state['messages'][0]['role'] == 'system':
            self.state['messages'][0]['content'] = prompt
        else:
            self.state['messages'].insert(0, {
                'role': 'system',
                'content': prompt
            })
        self.logger.info("System prompt updated")


# Example usage
if __name__ == '__main__':
    # Create and use the agent
    agent = ChatAgent({
        'system_prompt': 'You are a helpful coding assistant.',
        'max_history': 10
    })

    with agent:
        print("Chat Agent Example")
        print("=" * 50)

        # Simulate a conversation
        messages = [
            "Hello! Can you help me with Python?",
            "What's the difference between a list and a tuple?",
            "Thanks for the help!"
        ]

        for msg in messages:
            print(f"\nUser: {msg}")
            response = agent.execute(msg)
            print(f"Agent: {response}")

        print("\n" + "=" * 50)
        print(f"Total messages in history: {len(agent.get_history())}")
